partiql-lang/src/main/kotlin/org/partiql/lang/graph/GraphEngine.kt

```diff
@@ -0,0 +1,160 @@
+        fun restore(tree: StrideTree): List<ElemSpec> =
+            when (tree) {
+                is StrideLeaf -> tree.stride.elems
+                is StrideJoin -> {
+                    val left = restore(tree.left)
+                    val right = restore(tree.right)
+                    check(left.last() is NodeSpec)
+                    check(right.first() is NodeSpec)
+                    check(left.last() == right.first())
+                    left + right.tail
+                }
+            }
+        return StrideSpec(restore(strideTree))
+    }
+
+    /** Perform graph scans and joins, as the [StrideTree] plan specifies. */
+    private fun evaluatePlan(graph: Graph, plan: StrideTree): StrideResult {
+        return when (plan) {
+            is StrideLeaf -> {
+                val step = plan.stride.elems
+                check(step.size == 3, { "A leaf stride in a StrideTree plan must have exactly 3 elements" })
+                val lft = step[0] as NodeSpec // TODO: check these casts are ok
+                val edg = step[1] as EdgeSpec
+                val rgt = step[2] as NodeSpec
+                val stepSpec = StepSpec(edg.dir, Triple(lft.label, edg.label, rgt.label))
+                val triples = graph.getMatchingSteps(stepSpec)
+                StrideResult(
+                    plan.stride,
+                    triples.map { Stride(listOf(it.first, it.second, it.third)) }
+                )
+            }
+
+            is StrideJoin -> {
+                joinAdjacentStrides(
+                    evaluatePlan(graph, plan.left),
+                    evaluatePlan(graph, plan.right)
+                )
+            }
+        }
+    }
+
+    fun joinAdjacentStrides(left: StrideResult, right: StrideResult): StrideResult {
+        val leftSpec = left.spec.elems
+        val rightSpec = right.spec.elems
+        check(leftSpec.last() == rightSpec.first())
+        check(rightSpec.first() is NodeSpec)
+        val joinedSpec = leftSpec + rightSpec.tail
+
+        val joined = mutableListOf<Stride>()
+        for (lft in left.result) {
+            for (rgt in right.result) {
+                /** Relies on [Graph.Elem]s having proper equality */
+                if (lft.elems.last() == rgt.elems.first()) {
+                    joined.add(Stride(lft.elems + rgt.elems.tail))
+                }
+            }
+        }
+        return StrideResult(
+            StrideSpec(joinedSpec),
+            joined.toList()
+        )
+    }
+
+    /** Joins results of stride matches on distinct path patterns of a graph pattern.
+     *  In general, this is a cartesian product, but it is whittled down by the requirement
+     *  that a given binder variable binds to the same graph element in each individual answer.
+     */
+    fun joinStridesOnBinders(strides: List<StrideResult>): MatchResult {
+        return when (strides.size) {
+            0 -> { check(false); return /*dummy*/ MatchResult(emptyList(), emptyList()) }
```

Comment:
Is this check and dummy intended to essentially assert that this branch should be unreachable?

Reply:
yes